src/langchain_google_spanner/chat_message_history.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Cloud Spanner-based chat message history"""
+from __future__ import annotations
+
+from typing import List, Optional
+
+from google.cloud import spanner
+from google.cloud.spanner_admin_database_v1.types import DatabaseDialect  # type: ignore
+from google.cloud.spanner_v1 import param_types
+from google.cloud.spanner_v1.data_types import JsonObject
+from langchain_core.chat_history import BaseChatMessageHistory
+from langchain_core.messages import BaseMessage, messages_from_dict
+
```

Comment:
Since we are waiting on a Spanner review, can we also add the user agent in this PR or separately:
```suggestion
from .version import __version__

USER_AGENT = "langchain-google-spanner-python:chat_history" + __version__
```

Then update the logic for the client to append the user agent
https://github.com/googleapis/langchain-google-datastore-python/blob/3a62380794ad567ff8e9c0d5d7054093fb54cdb4/src/langchain_google_datastore/document_loader.py#L160-L168

Reply:
Added